crypto: Improve control of FIPS mode

stefanmb opened the pull request:
In Issue https://github.com/nodejs/node/issues/3819 requests were made to support a FIPS OpenSSL build of Node.js running in a non-FIPS mode. In PR https://github.com/nodejs/node/pull/3820 an attempt was made to introduce this feature.

In this PR I've attempted to merge all requests and discussion from https://github.com/nodejs/node/issues/3819 and https://github.com/nodejs/node/pull/3820.

The following features are introduced:

1. Default to FIPS off even in FIPS builds: FIPS mode now has to be explicitly enabled by command line argument, OpenSSL config file, or JavaScript API call.
2. Add JS API to check and control FIPS mode: Allow JavaScript applications to check if they are executing in FIPS mode and to enable/disable FIPS mode if applicable.
3. Add command line arguments to force FIPS on/off: --[disable|enable]-fips have been added, and they will override all other options.
4. Respect OPENSSL_CONF variable and read the config: OpenSSL provides a standard mechanism for reading a config file from the "OPENSSL_CONF" environment variable. Load this config file, which permits, among other things, to control FIPS mode.
5. Add testing for new features: Test combinations of features from 1 and 4.

As always, I'm open to any suggestions and improvements, especially if there is a better way to pass global options such as the enable/disable flags (instead of extern C variables).

These features were added as a result of discussion in https://github.com/nodejs/node/issues/3819 and https://github.com/nodejs/node/issues/3820, please refer to them for background information.

Note also that going forward we will need to run regression testing in FIPS builds twice: once with FIPS enabled at runtime, and once with FIPS disabled.

stefanmb commented:
@mhdawson @indutny @shigeki @jasnell @lordjabez @ScarletTanager This PR should include suggestions from all of you, any comments are greatly appreciated. Thanks!

mscdex commented:
Maybe defining a getter/setter simply called `fips` would be simpler instead of having separate methods?

mhdawson commented:
Added semver-major tag since it changes default behaviour

indutny commented on lib/crypto.js:
```diff
@@ -645,6 +647,13 @@ exports.getCurves = function() {
   return filterDuplicates(getCurves());
 };
 
+exports.hasFipsCrypto = function() {
```
Nit: `function hasFipsCrypto`

indutny commented on lib/crypto.js:
```diff
@@ -645,6 +647,13 @@ exports.getCurves = function() {
+  return hasFipsCrypto();
+};
+
+exports.setFipsCrypto = function(mode) {
```
Ditto.

indutny commented on src/node.cc:
```diff
@@ -3428,7 +3438,13 @@ static void ParseArgs(int* argc,
-#endif
+#if NODE_FIPS_MODE
+    } else if (strcmp(arg, "--disable-fips") == 0) {
+    disable_fips_crypto = true;
```
Nit: indent.

indutny commented on src/node.cc:
```diff
@@ -3428,7 +3438,13 @@ static void ParseArgs(int* argc,
+    } else if (strcmp(arg, "--disable-fips") == 0) {
+    disable_fips_crypto = true;
+    } else if (strcmp(arg, "--enable-fips") == 0) {
+    enable_fips_crypto = true;
```
Ditto.

indutny commented on src/node_crypto.cc:
```diff
@@ -5753,6 +5762,32 @@ void SetEngine(const FunctionCallbackInfo<Value>& args) {
+  bool mode = args[0]->BooleanValue();
+  if (disable_fips_crypto) {
+    return env->ThrowError(
+     "Cannot set FIPS mode, it was forced with --disable-fips at startup.");
```
Nit: indent, does `make cpplint` pass?

stefanmb replied:
Sadly yes, reports 0 errors found. I'll fix all this stuff.

indutny replied:
Yikes, too bad for it. Thank you! :+1:

indutny commented on src/node_crypto.cc:
```diff
@@ -5753,6 +5762,32 @@ void SetEngine(const FunctionCallbackInfo<Value>& args) {
+     "Cannot set FIPS mode, it was forced with --disable-fips at startup.");
+  } else if (enable_fips_crypto) {
+    return env->ThrowError(
+     "Cannot set FIPS mode, it was forced with --enable-fips at startup.");
```
Ditto.

indutny commented on test/parallel/test-crypto-authenticated.js:
```diff
@@ -93,7 +93,7 @@ for (var i in TEST_CASES) {
   (function() {
     if (!test.password) return;
-    if (common.hasFipsCrypto) {
+    if (crypto.hasFipsCrypto()) {
```
Let's make `common.hasFipsCrypto` a getter instead, should save us from fixing all tests and make them compatible to older node.js versions.

indutny commented on test/parallel/test-crypto-fips.js:
```diff
@@ -0,0 +1,194 @@
+}
+
+function getResponse(data)
+{
```
Nit: style, please put brace on the previous line. Does it pass `make lint`?

stefanmb replied:
It does :( I'll resolve the style problems.

indutny commented on test/parallel/test-crypto-fips.js:
```diff
@@ -0,0 +1,194 @@
+const FIPS_DISABLED = 0;
+const FIPS_ERROR_STRING = 'Error: Cannot set FIPS mode';
+const OPTION_ERROR_STRING = 'bad option';
+const CNF_FIPS_ON = common.fixturesDir + '/openssl_fips_enabled.cnf';
```
Nit: Let's use `path.join()` here.

indutny commented on test/parallel/test-crypto-fips.js:
```diff
@@ -0,0 +1,194 @@
+
+function getResponse(data)
+{
+  return data.toString().replace('\n', '').replace('>', '').trim();
```
`.replace(/\n|>/g, '').trim()`?

indutny commented on test/parallel/test-crypto-fips.js:
```diff
@@ -0,0 +1,194 @@
+}
+
+function childOk(child) {
+  console.log('Child ' +  ++num_children_ok + '/' + num_children_spawned +
```
It should be either `console.error` or nothing at all. Doesn't play well with TAP output.

indutny commented on test/parallel/test-crypto-fips.js:
```diff
@@ -0,0 +1,194 @@
+    env: env
+  });
+
+  console.log('Spawned child [pid:' + child.pid + '] with cmd ' +
```
Ditto.

indutny commented on test/parallel/test-crypto-fips.js:
```diff
@@ -0,0 +1,194 @@
+    child.stdin.setEncoding('utf-8');
+    child.stdout.on('data', function(data) {
+      // Prompt and newline may occur in undefined order.
+      const response = getResponse(data);
```
Perhaps it should wait until `end` event to accumulate all data? I'm afraid of the situation where `stdout` could be chunked.

indutny replied:
Also, this way it will be possible able to share lots of code between fips and non-fips branches in this function.

indutny commented on test/parallel/test-crypto-fips.js:
```diff
@@ -0,0 +1,194 @@
+      const response = getResponse(data);
+      if (response.length > 0) {
+        assert.notEqual(-1, response.indexOf(string));
+        childOk(child);
```
Possibly this will be called several times as mentioned above.

lordjabez commented:
This is a great change, and addresses all my concerns. Thanks!

indutny commented:
Overall I have really good impression from the PR. Thank you for your work!

mhdawson commented:
In terms of the command line options I'm wondering about the case were we want it enabled by default but then the application can turn on/off.   I also wonder about the need to prevent an app from turning it on. Maybe the following would make sense ?

--enable-fips  - enables FIPs, overrides setting in openssl file
--force-fips    - enables FIPs. overrides seeting in openss file, app cannot disable

stefanmb commented:
@mscdex I've switched to using a "crypto.fips" property as you suggested, thanks!
@indutny Thanks for the super quick feedback. I think I've addressed most of your concerns, I now use the 'end' event to validate output and 'data' event to buffer it. The handler is now shared between the FIPS and non-FIPS test branches. I'll double check the formatting issues.
@mhdawson I will remove --disable-fips and add a --force-fips tomorrow, as per our discussion there is no point in having disable-fips since that's the new default, instead we should provide a way to turn it on by default (--enable-fips) and a way to force it (--force-fips, does not allow it to be turned off).

indutny commented on src/node.cc:
```diff
@@ -3286,7 +3292,11 @@ static void PrintHelp() {
+#if NODE_FIPS_MODE
+         "  --disable-fips        force use of crypto without FIPS compliance"
+         "  --enable-fips         force use of crypto with FIPS compliance"
+#endif  /* NODE_FIPS_MODE */
```
I'm not sure about this one, but do we want to indent it @bnoordhuis ?

indutny commented on src/node.cc:
```diff
@@ -3428,7 +3438,13 @@ static void ParseArgs(int* argc,
+      disable_fips_crypto = true;
+    } else if (strcmp(arg, "--enable-fips") == 0) {
+      enable_fips_crypto = true;
+#endif /* NODE_FIPS_MODE */
```
Ditto, @bnoordhuis

indutny commented on src/node_crypto.cc:
```diff
@@ -3871,24 +3873,26 @@ SignBase::Error Sign::SignFinal(const char* key_pem,
-      goto exit;
+  if (FIPS_mode()) {
+    /* Validate DSA2 parameters from FIPS 186-4 */
+    if (EVP_PKEY_DSA == pkey->type) {
```
Just for the sake of saving `git blame` data. Could we merge it into one `if` with `&&`?

indutny commented on test/parallel/test-crypto-fips.js:
```diff
@@ -0,0 +1,208 @@
+      cmd + ' and args \'' + args + '\'');
+  num_children_spawned++;
+
+  const response_handler = function(response, expectedOutput, expectedError) {
```
Just curious, any point in doing it this way, instead of just `function responseHandler()`?

indutny commented on test/parallel/test-crypto-fips.js:
```diff
@@ -0,0 +1,208 @@
+    // We won't always have data on both stdout and stderr.
+    if (undefined !== response) {
+      if (EXIT_FAILURE === expectedOutput) {
+        if(-1 === response.indexOf(expectedError))
```
Linting :)

indutny commented on test/parallel/test-crypto-fips.js:
```diff
@@ -0,0 +1,208 @@
+    // Prompt and newline may occur in undefined order.
+    const response = data.toString().replace(/\n|>/g, '').trim();
+    if (response.length > 0) {
+      child_responses[child] = response;
```
Should probably do `+= ` here, and initialize it to `''` initially.

stefanmb replied:
Good catch, I meant to do that! Thanks, will be fixed soon.

indutny commented on test/parallel/test-crypto-fips.js:
```diff
@@ -0,0 +1,208 @@
+const CNF_FIPS_OFF = path.join(common.fixturesDir, 'openssl_fips_disabled.cnf');
+var num_children_spawned = 0;
+var num_children_ok = 0;
+const child_responses = {};
```
Is there any point in declaring it this way? I think it could live below in `testHelper`'s closure. Also it could be just two vars `stdout = ''` and `stderr = ''`.

stefanmb commented:
@indutny Thanks. Should be better now, apologies if the JavaScript is not too idiomatic. I still wrapped the buffer strings in an object so I can access them 'by reference'.

indutny commented on test/parallel/test-crypto-fips.js:
```diff
@@ -0,0 +1,208 @@
+  }
+
+  // Buffer all data received from children on stderr/stdout.
+  const buffer = { stderr: '', stdout: '' };
```
Let's do it this way:

```js
var stderr = '';
var stdout = '';
```

stefanmb replied:
Perhaps a silly question - but if I do this, in the response_buffer function how do I differentiate between the two variables without a branch and text comparison on the stream's string value? Right now I can simply do "buffer[stream]" instead of checking which variable I write to based on the value of stream. I apologize if I am missing something obvious.

indutny replied:
See comments below, you don't need `responseBuffer` at all. Just inline `stdout += `, `stderr +=` into event listeners.

indutny commented on test/parallel/test-crypto-fips.js:
```diff
@@ -0,0 +1,208 @@
+  const buffer = { stderr: '', stdout: '' };
+  const response_buffer = function(stream, data) {
+    // Prompt and newline may occur in undefined order.
+    const response = data.toString().replace(/\n|>/g, '').trim();
```
Let's put this into a `filter` function:

```js
function filter(data) {
  return data.toString().replace(/\n|>/g, '').trim();
}
```

indutny commented on test/parallel/test-crypto-fips.js:
```diff
@@ -0,0 +1,208 @@
+  };
+
+  child.stdout.on('data', function(data) {
+    response_buffer('stdout', data);
```
Let's do `stdout += filter(data);` here.

indutny commented on test/parallel/test-crypto-fips.js:
```diff
@@ -0,0 +1,208 @@
+  });
+
+  child.stderr.on('data', function(data) {
+    response_buffer('stderr', data);
```
`stderr += filter(data);` here.

indutny commented on test/parallel/test-crypto-fips.js:
```diff
@@ -0,0 +1,208 @@
+      cmd + ' and args \'' + args + '\'');
+  num_children_spawned++;
+
+  function response_handler(response, expectedOutput, expectedError) {
```
`responseHandler`, please.

indutny commented on test/parallel/test-crypto-fips.js:
```diff
@@ -0,0 +1,208 @@
+  // If using FIPS mode binary, or running a generic test.
+  if (compiledWithFips() || !requiresFips) {
+    child.stdout.on('end', function(data) {
+      response_handler(buffer['stdout'],
```
Just `stdout` here.

indutny commented on test/parallel/test-crypto-fips.js:
```diff
@@ -0,0 +1,208 @@
+  } else {
+    // If using a non-FIPS binary expect a failure.
+    child.stdout.on('end', function() {
+      response_handler(buffer['stdout'],
```
Ditto.

indutny commented on test/parallel/test-crypto-fips.js:
```diff
@@ -0,0 +1,208 @@
+    });
+    /* Failure to start Node executable is reported on stderr */
+    child.stderr.on('end', function() {
+      response_handler(buffer['stderr'],
```
`stderr`.

indutny commented on test/parallel/test-crypto-fips.js:
```diff
@@ -0,0 +1,208 @@
+  // first, resulting in EPIPE. We can avoid EPIPE by  not writing anything.
+  if (null !== cmd) {
+    child.stdin.setEncoding('utf-8');
+    child.stdin.write(cmd + '\n;process.exit()\n');
```
Just curious... could we possibly start child process with `-e` command-line argument instead? And maybe use `child_process.spawnSync` to simplify all of the code above?

jelmd commented:
Can you guarantee, that FIPS_mode() got called before using RAND_* functions are used, e.g. in EntropySource()/CheckEntropy()? At least in 5.0..1.x on Solaris with a FIPS capable lib and FIPS mode disabled, this was the cause of hangings if not done before ...

stefanmb commented:
@jelmd I recall seeing a note about that in a patch you posted here http://iws.cs.uni-magdeburg.de/~elkner/tmp/node5/ssl.patch. I'll review your change. Sorry for taking a while with finishing this, I'm caught up in a few other things. Thanks for the patience.

jelmd commented:
OK, thanx a lot, too! :)

chrisamoore commented:
/cc @lordjabez

stefanmb commented:
@indutny I've rewritten the tests to use spawnSync which makes them simpler, as you indicated. I've also removed --disable-fips and and added --force-fips as per @mhdawson's suggestion. I think that makes sense given that the default is now FIPS off. Thanks again for your help, the code is more readable now. Also, "make (js)lint" passes. :)

@jelmd I don't have access to a Solaris machine, but I looked again at your patch and I don't think it's the call to FIPS_mode() that you require, but rather the underlying OPENSSLL_init() call. I cannot test that this fixes the issue you were experiencing, but it strikes me as something we should be doing anyway before setting V8's entropy source to use OpenSSL. Using the FIPS canister changes OpenSSL's behaviour with respect to the RAND_* family functions. Note that OPENSSL_init() is always called by FIPS_mode(), meaning that behaviour between a FIPS build with FIPS mode off and a non-FIPS build may not be identical. This appears to be how OpenSSL works by design. Any comments on this are welcome @indutny.

Thanks again everyone!

indutny commented on src/node.cc:
```diff
@@ -3282,7 +3288,11 @@ static void PrintHelp() {
          "  --tls-cipher-list=val use an alternative default TLS cipher list\n"
-#endif
+#if NODE_FIPS_MODE
+         "  --disable-fips        force use of crypto without FIPS compliance"
```
Stale docs.

indutny commented on src/node_crypto.cc:
```diff
@@ -3858,7 +3860,7 @@ SignBase::Error Sign::SignFinal(const char* key_pem,
 #ifdef NODE_FIPS_MODE
   /* Validate DSA2 parameters from FIPS 186-4 */
-  if (EVP_PKEY_DSA == pkey->type) {
+  if (FIPS_mode() && (EVP_PKEY_DSA == pkey->type)) {
```
Any need in parens here?

indutny commented:
One comment, otherwise LGTM

jelmd commented:
@stefanmb: Wrt. FIPS_mode()/OpenSSL_init(): yepp, probably. But since I was not sure about all the code pathes leading to EntropySource(..), IIRC the idea was: we need to check/set the FIPS_mode() at this place anyway, so digging deeper into FIPS_mode() doesn't buy much. At least the fact, that it hang in a test implied to me, that  InitCryptoOnce() (where the FIPS_mode() call get made) either isn't in its code path _OR_ the setting happens too late, i.e. one of the other functions called before in InitCryptoOnce() involve EntropySource(...), which than in turn means, it might be used not in the intended FIPS_mode()...

Makes sense?

stefanmb commented:
@indutny Thanks! Fixed - changed the descriptions of cmd line args to better match functionality, removed the extra parenthesis.

stefanmb commented:
@jelmd I understand what you're saying. I think that the call to FIPS_mode in InitCryptoOnce was indeed too late, but I do not want to move the initialization earlier because it's closely related to loading the crypto module (e.g. it uses libuv primitives to insure single initialization) and changing the startup path for the crypto module is a very high risk move. Instead, I think adding the OpenSSL_init call as I did will achieve the same results for you with minimal risk. I also think calling OpenSSL_init directly is more obvious to the reader, rather than having it be called as a side-effect of FIPS_mode().

WIth respect to naming, I agree that getFipsCrypto makes more sense (as a counterpart to setFipsCrypto) so I have changed it accordingly. The old naming is based on properties in test/common,js (hasCrypto/hasFipsCrypto). Note that that these functions are only exposed to JavaScript through the crypto.fips property anyway.

With respect to the #ifdef NODE_FIPS_MODE and failing early, I'm not sure I understand. When compiling without FIPS the call to OpeSSL_init (or FIPS_mode) won't do anything anyway, so there's no point in making it (the relevant code is ifdef'd out within the OpenSSL source).

indutny commented on src/node.cc:
```diff
@@ -3282,7 +3288,11 @@ static void PrintHelp() {
          "  --tls-cipher-list=val use an alternative default TLS cipher list\n"
-#endif
+#if NODE_FIPS_MODE
+         "  --enable-fips         enable FIPS compliant crypto on startup"
```
Missing `\n` at the end of the lines. Sorry!

indutny commented:
One last nit, sorry!

stefanmb commented:
@indutny Fixed the newline! Yikes!

indutny commented:
LGTM, but I would like to also have one more LGTM from @nodejs/crypto or @mhdawson . Thank you!

indutny commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/1687/

indutny commented:
Another CI run: https://ci.nodejs.org/job/node-test-pull-request/1689/ . There was some compilation problem on Windows.

jelmd commented:
@stefanmb: wrt. changing the startup path: IC. Using OpenSSL_init() instead of FIPS_mode() seems to be ok. My point was just, that the current FIPS_mode() may have an influence on what is \[not\] allowed or made, when RAND_* gets called. Therefore the safe way is IMHO to check and set the FIPS_mode before calling RAND_*...

getFipsCrypto: thanx :). And if one compiles w/o fips support, crypto.fips is undefined, right?

"fail early": I'm programming a little bit defensive and thus try to avoid digging into implementation details/making pre-optimizations. So if the doc e.g. says, "must be called before any other action takes place", I try to adhere if it isn't too costly, because implementation may change. This way I may waste some cpu cycles but on the other side I do not need to monitor every change in the lib or don't get hit later by accident because a more or less noop call wasn't made. Last but not least OpenSSL_init() and the notes sound important/somehow imply to me, that if this call gets through most of the SSL stuff will work, otherwise all bets are open .... So I would call it always at the start of an app to have it fail early and not when SSL is actually needed. Yepp, depending on what the app is going to do, the lazy way might be ok as well, e.g. if SSL is probably not needed at all in the app, or one is able to recover from such errors easily without huge headaches ...

Anyway, when you say OpenSSL_init() is a noop in non-fips capable linkage and the FIPS_mode has no influence on RAND_* (and if eventually changed in the future, that one definitely takes care of it) I would rely on this ;-)

mhdawson commented:
@indutny Will review later today

mhdawson commented:
LGTM from me, but before we land I want to understand how we need to change the CI job to make sure it continues to validate whats needed. @indunty I can co-ordinate with @stefanmb on that front and then land this PR and change the build jobs at the same time.

stefanmb commented:
@mhdawson @indutny I've updated this PR to resolve issues encountered while attempting to setup the CI:

1) test/common.js still used the stale API (crypto.hasFipsCrypto instead of fips.crypto)

2) I've updated the testing to survive a polluted environment (i.e. one where OPENSSL_CONF may already be defined, note that OpenSSL will not allow you to set the FIPS mode to on if it's already on, an exception will be thrown).

3) During CI for a FIPS build we must now run the test suite twice, once with FIPS enabled and once with FIPS disabled. One way to do so is to use the OPENSSL_CONF environment variable, however this will cause test failures for those tests that use opensslCli from test/common.js because the OPENSSL_CONF variable will affect the standalone OpenSSL utilities. We could filter out the OPENSSL_CONF variable from the environment in common.js, but this does not strike me as a clean solution. The alternative is to pass "--enable-fips" to each test, however there was no obvious way to pass the --enable-fips argument through the "make test-ci" target. The "--special-command" argument comes close, but is only able to suffix or prefix the entire invocation. I have made a separate PR to deal with this problem since it outside the scope of this PR: https://github.com/nodejs/node/pull/5376

mhdawson commented:
LTGM to latest changes, and successful test run in both fiips/non-fips enabled in updated test here:  https://ci.nodejs.org/job/node-test-commit-linux-fips-dawson/

Will sync landing this and updating the test job.

One last regular CI run so that we have a current one: https://ci.nodejs.org/job/node-test-pull-request/1752/

mhdawson commented:
CI run is green, landing

mhdawson commented:
Landed as 7c48cb560108bc2f4365bb0c290a9cd68dcc56e4

Updated fits test, CI run here to validate test is updated ok: https://ci.nodejs.org/job/node-test-commit/2345/

mhdawson commented:
CI all good closing
